### PR TITLE
PAYARA-1909 Domain Discovery Mode

### PR DESCRIPTION
**Pull request by smillidge**
Major Refactor of the Hazlecast Integration. By default Hazlecast now uses information in the domain.xml to determine where members of the cluster are likely to be. This results in pretty transparent clustering of all instances in the domain.
All instances and the DAS now form a single unified Hazelcast Data Grid for data sharing including web session data and JCache data.
Tested on Azure in a VPC all instances in the domain clustered automatically using SSH nodes
Tested on AWS in a VPC all instances in the domain clustered automatically using SSH nodes
Tested on Google Compute in a VPC all instances in the domain clustered automatically using SSH nodes
Tested on LAN in Linux all instances in the domain clustered automatically using SSH nodes

Payara Micro has new flag --clustermode which can be domain:host:port to cluster with a DAS (or any other instance) or it can be tcpip:host:port:host:port to provide a list of hosts and ports to check. This change make it easy to cluster micros in the cloud where IPs are known e.g. VPC. Default is the original multicast

Host aware partitioning is now the default.

### CODE REVIEW

**Comment by smillidge** [comment]
Jenkins quick test Payara 5 please

**Comment by smillidge** [comment]
Jenkins quick test Payara 5 please

File: appserver/extras/payara-micro/payara-micro-core/src/main/java/fish/payara/micro/impl/PayaraMicroImpl.java [hunk 1]
```diff
@@ -1960,7 +2000,7 @@ private void setArgumentsFromSystemProperties() {
         sslPort = getIntegerProperty("payaramicro.sslPort", Integer.MIN_VALUE);
         hzMulticastGroup = getProperty("payaramicro.mcAddress");
         hzPort = getIntegerProperty("payaramicro.mcPort", Integer.MIN_VALUE);
-        hostAware = getBooleanProperty("payaramicro.hostAware");
+        hostAware = getBooleanProperty("payaramicro.hostAware","true");
```
Comment (arjantijms) [inline appserver/extras/payara-micro/payara-micro-core/src/main/java/fish/payara/micro/impl/PayaraMicroImpl.java]: >hostAware = getBooleanProperty("payaramicro.hostAware","true");

Could do with a space after the ,

File: appserver/extras/payara-micro/payara-micro-core/src/main/java/fish/payara/micro/impl/PayaraMicroImpl.java [hunk 2]
```diff
@@ -2341,6 +2391,19 @@ private Boolean getBooleanProperty(String value) {
         }
         return "true".equals(property);
     }
+    
+    private Boolean getBooleanProperty(String value, String defaultValue) {
+        String property;
```
Comment (arjantijms) [inline appserver/extras/payara-micro/payara-micro-core/src/main/java/fish/payara/micro/impl/PayaraMicroImpl.java]: `property = System.getProperty(value);`

could become

`String property = System.getProperty(value);`

Why separate those?

File: nucleus/admin/rest/rest-service/src/main/java/org/glassfish/admin/rest/utils/JsonUtil.java [hunk 3]
```diff
@@ -140,7 +141,9 @@ public static JsonValue getJsonValue(Object object, boolean hideConfidentialProp
             result = Json.createValue((BigDecimal) object);
         } else if (object.getClass().isEnum()){
             result = Json.createValue(object.toString());
-        } else {
+        } else if (object instanceof InetAddress) {
+            result = Json.createValue(object.toString());
+        }else {
```
Comment (arjantijms) [inline nucleus/admin/rest/rest-service/src/main/java/org/glassfish/admin/rest/utils/JsonUtil.java]: `}else {`

Could use a space between after {

File: nucleus/payara-modules/hazelcast-bootstrap/src/main/java/fish/payara/nucleus/hazelcast/MemberAddressPicker.java [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+ * @author Steve Millidge (Payara Foundation)
+ */
+public class MemberAddressPicker implements MemberAddressProvider {
+
+    private final ServerEnvironment env;
+    private final HazelcastRuntimeConfiguration config;
+    private final HazelcastConfigSpecificConfiguration localConfig;
+    private InetSocketAddress bindAddress;
+    private InetSocketAddress publicAddress;
+    private static final Logger logger = Logger.getLogger(MemberAddressPicker.class.getName());
```
Comment (arjantijms) [inline nucleus/payara-modules/hazelcast-bootstrap/src/main/java/fish/payara/nucleus/hazelcast/MemberAddressPicker.java]: By convention, static variables are put before instance variables, so the logger could be moved up to line 61

File: nucleus/payara-modules/hazelcast-bootstrap/src/main/java/fish/payara/nucleus/hazelcast/MemberAddressPicker.java [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+    @Override
+    public InetSocketAddress getBindAddress() {
+        return bindAddress;
+    }
+
+    @Override
+    public InetSocketAddress getPublicAddress() {
+        if (publicAddress != null) {
+            return publicAddress;
+        } else {
```
Comment (arjantijms) [inline nucleus/payara-modules/hazelcast-bootstrap/src/main/java/fish/payara/nucleus/hazelcast/MemberAddressPicker.java]: It's a matter of style, but the else is not needed here. Can return directly after the if block.

**Comment by smillidge** [comment]
Jenkins quick test Payara 5 please

**Comment by lprimak** [comment]
Agreed with @arjantijms style comments above, although all are minor

**Comment by lprimak** [comment]
GUI UX usability issue:
When on the Data Grid configuration screen, it has a radio box: multicast/tcpip/domain
The issue is that parameters for all three options are present below, which is confusing.
The parameters should only be there for applicable setting of the radio button, and change depending which radio button is selected.
I think this is implemented somewhere else in the admin console as well, I don't remember where though.

**Comment by lprimak** [comment]
Ran a simple test: 1 DAS and 1 Micro.  No changes (default domain1 configuration)
Got the following error from Micro:

```
[2017-12-07T00:08:35.605-0600] [] [SEVERE] [] [com.hazelcast.cluster] [tid: _ThreadID=42 _ThreadName=hz._hzInstance_1_development.priority-generic-operation.thread-0] [timeMillis: 1512626915605] [levelValue: 1000] [10.0.1.15]:5900 [development] [3.9] Node could not join cluster. A Configuration mismatch was detected: Incompatible joiners! expected: tcp-ip, found: multicast Node is going to shutdown now!
```

With default options, Micro should start correctly, otherwise it breaks the "works out-of-the-box" principle

**Comment by smillidge** [comment]
That will require chanting micro default cluster mode to domain. However micro is a standalone product and clusters by multicast by default. As your tested scenario is not the default usage model for micro I disagree about it working out of the box.

**Comment by smillidge** [comment]
Hazelcast 3.9 still needs all the interface checking unless it has changed in a later minor release

**Comment by smillidge** [comment]
Ah I see what you mean. Payara Micro does not start if a DAS is running. I'll try and work out how it is trying to join the DAS in a cluster.

**Comment by smillidge** [comment]
I'll investigate and recreate the PR

**Pull request closed without merging**